DirectProgramming/DPC++FPGA/Tutorials/Features/ac_int/src/ac_int.cpp

```diff
@@ -0,0 +1,236 @@
+//==============================================================
+// Copyright Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+#include <CL/sycl.hpp>
+#include <sycl/ext/intel/ac_types/ac_int.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+
+#include <bitset>
+
+// dpc_common.hpp can be found in the dev-utilities include folder.
+// e.g., $ONEAPI_ROOT/dev-utilities/include/dpc_common.hpp
+#include "dpc_common.hpp"
+
+using namespace sycl;
+
+// Forward declare the kernel names in the global scope.
+// This FPGA best practice reduces name mangling in the optimization reports.
+class BasicOpsInt;
+class BasicOpsAcInt;
+class ShiftOp;
+class EfficientShiftOp;
+class BitOps;
+
+using MyUInt2 = ac_int<2, false>;
+using MyInt14 = ac_int<14, true>;
+using MyInt15 = ac_int<15, true>;
+using MyInt28 = ac_int<28, true>;
+
+void TestBasicOpsInt(queue &q, const int &a, const int &b, int &c, int &d,
```

Comment:
Why do these test functions all use references to `a`, `b`, etc instead of passing by value? I don't remember seeing that as a best practice anywhere.

Reply:
No particular reason. Just because that's a general best practice for passing arguments in C++.

Reply:
Agreed^

Reply:
I disagree. [Always passing by reference is silly](https://stackoverflow.com/questions/2627166/difference-between-const-reference-and-normal-parameter).

1. pass by value when you want to pass a small primitive (such as `int`) and don't want it to get modified
2. pass by reference when you want to pass a large struct, or when you want the function to modify it
3. pass a pointer when you are a C programmer pretending to know C++ ;) or rather, when you want to assign null, or do pointer arithmetic or any of those other techniques that are not possible with references.

Since this function only modifies `c`, `d`, and `e`, I think the signature should look like this:

```suggestion
void TestBasicOpsInt(queue &q, int a, int b, int &c, int &d, int &e)
```

Reply:
Not worth arguing about, your choice @shuoniu-intel